src/librustc_resolve/def_collector.rs

```diff
@@ -117,10 +118,8 @@ impl<'a> visit::Visitor<'a> for DefCollector<'a> {
                 // we must mirror everything that `visit::walk_fn` below does.
                 self.visit_fn_header(&sig.header);
                 visit::walk_fn_decl(self, &sig.decl);
-                if let Some(body) = body {
-                    let closure_def = self.create_def(closure_id, DefPathData::ClosureExpr, span);
-                    self.with_parent(closure_def, |this| this.visit_block(body));
-                }
+                let closure_def = self.create_def(closure_id, DefPathData::ClosureExpr, span);
+                self.with_parent(closure_def, |this| walk_list!(this, visit_block, body));
```

Comment:
I think it's a problem in item lowering rather than here?
If there's no body then there's no closure, but `async fn` lowering still wraps the `ExprKind::Err` recovered from the "no block" into a closure instead of just returning an `ExprKind::Err`.

Reply:
I looked a bit now. Changing lowering to avoid creating the closure on `body = None` would also work I think, but it seems to me it would duplicate more logic and break the `fn lower_block_expr_opt` abstraction used in a bunch of places. Treating `qualifiers fn foo();` as `qualifiers fn foo() { err }` seems fairly consistent and it also simplifies the logic in `def_collector` as well.

Reply:
Ok.